### PR TITLE
Binary classification samples update

### PR DESCRIPTION
**Pull request by artidoro**
Tracked in #2522.

In this PR I use the templates for binary classification samples to eliminate the dependency on the SamplesUtils and make the samples standalone.

Some binary classification trainers did not have samples, so I added them.

Note: This PR does not take care of FFM on which @wschin will be working.

### CODE REVIEW

**Comment by artidoro** [comment]
Noticed that I forgot to add `Gam` and `LinearSVM`.

File: docs/samples/Microsoft.ML.Samples/Program.cs [hunk 1]
```diff
@@ -1,12 +1,36 @@
...
+using Samples.Dynamic.Trainers.BinaryClassification;
 
 namespace Microsoft.ML.Samples
 {
     internal static class Program
     {
         static void Main(string[] args)
         {
-            CalculateFeatureContribution.Example();
+            Console.WriteLine("Prior");
```
Comment (artidoro) [inline docs/samples/Microsoft.ML.Samples/Program.cs]: Drop changes to Program.cs.

**Comment by shmoradims** [comment]
> using Samples.Dynamic;

this file shouldn't be in the PR. let's remove it to reduce the chance of conflicts.

---
Refers to: docs/samples/Microsoft.ML.Samples/Program.cs:1 in ace5d1b. [](commit_id = ace5d1b81e2fee0e7092663276a0074e13049f7a, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/BinaryClassification.ttinclude [hunk 2]
```diff
@@ -56,7 +64,13 @@ namespace Samples.Dynamic.Trainers.BinaryClassification
...
             // Evaluate the overall metrics
             var metrics = mlContext.BinaryClassification.<#=Evaluator#>(transformedTestData);
-            Microsoft.ML.SamplesUtils.ConsoleUtils.PrintMetrics(metrics);
+            Console.WriteLine($"Accuracy: {metrics.Accuracy:F2}");
+            Console.WriteLine($"AUC: {metrics.AreaUnderRocCurve:F2}");
+            Console.WriteLine($"F1 Score: {metrics.F1Score:F2}");
+            Console.WriteLine($"Negative Precision: {metrics.NegativePrecision:F2}");
+            Console.WriteLine($"Negative Recall: {metrics.NegativeRecall:F2}");
+            Console.WriteLine($"Positive Precision: {metrics.PositivePrecision:F2}");
+            Console.WriteLine($"Positive Recall: {metrics.PositiveRecall:F2}");
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/BinaryClassification.ttinclude]: thanks for changing this. I think, it would be cleaner to replicate ConsoleUtils.PrintMetrics as a helper function PrintMetrics at the bottom of the .ttinclude file, so that Example() becomes leaner.

**Comment by shmoradims** [comment]
> {

all checklist points are done. please also add #6:
Inspect main properties (weights, splits, etc.) of the model. Here are examples for trees, GAM, linear. Simple is enough here. We just want to give users some ideas about what's possible. Only inspect 2-3 main model properties (not all). Also, no need to have a for-loop here. For example, if a tree model has 100 trees, take the first tree and print 2-3 main properties of that tree (NumberOfLeaves, NumberOfNodes).

---
Refers to: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/BinaryClassification.ttinclude:11 in ace5d1b. [](commit_id = ace5d1b81e2fee0e7092663276a0074e13049f7a, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/PriorTrainer.cs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            Console.WriteLine($"AUC: {metrics.AreaUnderRocCurve:F2}");
+            Console.WriteLine($"F1 Score: {metrics.F1Score:F2}");
+            Console.WriteLine($"Negative Precision: {metrics.NegativePrecision:F2}");
+            Console.WriteLine($"Negative Recall: {metrics.NegativeRecall:F2}");
+            Console.WriteLine($"Positive Precision: {metrics.PositivePrecision:F2}");
+            Console.WriteLine($"Positive Recall: {metrics.PositiveRecall:F2}");
+            
+            // Expected output:
+            //   Accuracy: 0.52
+            //   AUC: 0.50
```
Comment (shmoradims) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/PriorTrainer.cs]: >0.50 [](start = 22, length = 4)

AUC .5 is the worst possible (same as random coin flip). can we make it any better?

**Comment by shmoradims** [comment]
please also exclude libmf change in this PR

**Comment by shmoradims** [comment]
i like this PR very much :)

**Comment by artidoro** [comment]
Add reference to:
```csharp
        // This example requires installation of additional nuget package <a href="https://www.nuget.org/packages/Microsoft.ML.LightGbm/">Microsoft.ML.LightGbm</a>.
```
For `LightGbm` extensions.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/BinaryClassification.ttinclude [hunk 4]
```diff
@@ -23,6 +23,14 @@ namespace Samples.Dynamic.Trainers.BinaryClassification
 
             // Convert the list of data points to an IDataView object, which is consumable by ML.NET API.
             var trainingData = mlContext.Data.LoadFromEnumerable(dataPoints);
+<# if (CacheData) { #>
+
+            // ML.NET doesn't cache data set by default. Therefore, if one reads a data set from a file and accesses it many times,
+			// it can be slow due to expensive featurization and disk operations. When the considered data can fit into memory,
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/BinaryClassification.ttinclude]: >			 [](start = 0, length = 3)

tabs? HERESY!

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/LightGbm.cs [hunk 5]
```diff
@@ -1,41 +1,104 @@
...
+using System.Linq;
+using Microsoft.ML;
+using Microsoft.ML.Data;
 
 namespace Samples.Dynamic.Trainers.BinaryClassification
 {
-    public class LightGbm
+    public static class LightGbm
     {
-        // This example requires installation of additional nuget package <a href="https://www.nuget.org/packages/Microsoft.ML.LightGbm/">Microsoft.ML.LightGbm</a>.
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/LightGbm.cs]: >a href="https://www.nuget.org/packages/Microsoft.ML.LightGbm/">Microsoft.ML.LightGbm</a> [](start = 75, length = 88)

can we preserve this line?

Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/LightGbm.cs]: Yes thank you for pointing it out I am adding it back now :)

**Comment by Ivanidzo4ka** [comment]
> // Licensed to the .NET Foundation under one or more agreements.

What is going on with `libmf`  folder? why you editing it?

---
Refers to: src/Microsoft.ML.StandardTrainers/StandardTrainersCatalog.cs:1 in ace5d1b. [](commit_id = ace5d1b81e2fee0e7092663276a0074e13049f7a, deletion_comment = False)

File: src/Microsoft.ML.Mkl.Components/MklComponentsCatalog.cs [hunk 6]
```diff
@@ -80,7 +80,7 @@ public static class MklComponentsCatalog
         /// <example>
         /// <format type="text/markdown">
         /// <![CDATA[
-        /// [!code-csharp[SymbolicStochasticGradientDescent](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/SymbolicStochasticGradientDescent.cs)]
+        /// [!code-csharp[SymbolicStochasticGradientDescent](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/SymbolicSgdLogisticRegression.cs)]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Mkl.Components/MklComponentsCatalog.cs]: >SymbolicStochasticGradientDescent [](start = 26, length = 33)

does it make sense to align with name of the method?

File: src/Microsoft.ML.StandardTrainers/StandardTrainersCatalog.cs [hunk 7]
```diff
@@ -269,6 +269,12 @@ public static class StandardTrainersCatalog
         /// </summary>
         /// <param name="catalog">The binary classification catalog trainer object.</param>
         /// <param name="options">Trainer options.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        ///  [!code-csharp[SDCA](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/SdcaNonCalibratedWithOptions.cs)]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardTrainers/StandardTrainersCatalog.cs]: >SDCA [](start = 27, length = 4)

`SdcaNonCalibrated` align with method name?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/LinearSvm.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            Console.WriteLine($"Accuracy: {metrics.Accuracy:F2}");
+            Console.WriteLine($"AUC: {metrics.AreaUnderRocCurve:F2}");
+            Console.WriteLine($"F1 Score: {metrics.F1Score:F2}");
+            Console.WriteLine($"Negative Precision: {metrics.NegativePrecision:F2}");
+            Console.WriteLine($"Negative Recall: {metrics.NegativeRecall:F2}");
+            Console.WriteLine($"Positive Precision: {metrics.PositivePrecision:F2}");
+            Console.WriteLine($"Positive Recall: {metrics.PositiveRecall:F2}");
+            
+            // Expected output:
+            //   Accuracy: 0.53
```
Comment (Ivanidzo4ka) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/LinearSvm.cs]: >0.53 [](start = 27, length = 4)

why it so atrociously bad? Would increase in learning rate and num of iteration improve it?

**Comment by Ivanidzo4ka** [comment]
>             // Look at 5 predictions

Print maybe?

---
Refers to: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/BinaryClassification.ttinclude:58 in ace5d1b. [](commit_id = ace5d1b81e2fee0e7092663276a0074e13049f7a, deletion_comment = False)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/PriorTrainer.cs (hunk 3 above)
Comment (shmoradims) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/PriorTrainer.cs]: it's still 0.5.

---
In reply to: [275048035](https://github.com/dotnet/machinelearning/pull/3311#discussion_r275048035) [](ancestors = 275048035)

Comment (artidoro) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/BinaryClassification/PriorTrainer.cs]: So I changed the way we generate the data, so that the Accuracy is greater than 0.5. However, for Prior trainer the AUC will always be 0.50. The reason is that there are only two points in the ROC curve: (0,0) and (1,1) one for when the threshold is higher than the prior probability, and one for when it is lower. That creates a line graph with area 0.50. There is no way to increase it for prior.
